File: tools/tf-types-ts/src/core/canonical.ts
/**
 * Deterministic JSON serialization compatible with the Rust implementation.
 *
 * Rules:
 *   - Object keys are sorted lexicographically by their NFC-normalized
 *     codepoint sequence.
 *   - All string values are NFC-normalized.
 *   - Finite integers emit as integers (no ".0"); finite non-integer numbers
 *     emit via JavaScript's shortest round-trip representation.
 *   - -0 is emitted as 0.
 *   - undefined, functions, symbols, NaN, ±Infinity are rejected.
 *   - No whitespace anywhere in the output.
 *
 * The output is a valid UTF-8 JSON string; byte-for-byte equality with the
 * Rust implementation is tested by canonical-vectors.yaml.
 */
export function canonicalize(value: unknown): string {
  return encode(value);
}

export class CanonicalJsonError extends Error {}

function encode(v: unknown): string {
  if (v === null) return "null";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") return encodeNumber(v);
  if (typeof v === "bigint") return v.toString(10);
  if (typeof v === "string") return JSON.stringify(v.normalize("NFC"));
  if (Array.isArray(v)) return "[" + v.map(encode).join(",") + "]";
  if (typeof v === "object") {
    const entries = Object.entries(v as Record<string, unknown>)
      .filter(([, val]) => val !== undefined)
      .map(([k, val]) => [k.normalize("NFC"), val] as const)
      .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
    return "{" + entries.map(([k, val]) => JSON.stringify(k) + ":" + encode(val)).join(",") + "}";
  }
  throw new CanonicalJsonError(`cannot canonicalize value of type ${typeof v}`);
}

function encodeNumber(n: number): string {
  if (!Number.isFinite(n)) throw new CanonicalJsonError(`cannot canonicalize non-finite number: ${n}`);
  if (Object.is(n, -0)) return "0";
  if (Number.isInteger(n)) return n.toFixed(0);
  return String(n);
}

File: crates/tf-types/src/canonical.rs
//! Deterministic JSON serialization compatible with the TypeScript
//! implementation in `tools/tf-types-ts/src/core/canonical.ts`.
//!
//! Rules:
//!   * Object keys are sorted by UTF-8 byte order of their NFC-normalized
//!     form. (Rust's `String::cmp` compares underlying UTF-8 bytes; the
//!     TS implementation uses an explicit UTF-8 byte comparator instead
//!     of JS's UTF-16 code-unit `<`.)
//!   * All string values are NFC-normalized.
//!   * Finite integers emit as integers (no `.0`); finite non-integer numbers
//!     emit via Rust's shortest round-trip representation, matching the
//!     `serde_json` defaults which in turn match JavaScript's `String(n)`.
//!   * `-0` is emitted as `0`.
//!   * `NaN`, `±Infinity` are rejected.
//!   * No whitespace anywhere in the output.
//!
//! Byte-for-byte parity with the TypeScript implementation is enforced by
//! `conformance/canonical-vectors.yaml` and
//! `conformance/cross-language-signature-vectors.yaml`.

use std::fmt::Write;

use serde_json::{Map, Value};

#[derive(Debug, thiserror::Error)]
pub enum CanonicalJsonError {
    #[error("cannot canonicalize non-finite number: {0}")]
    NonFinite(f64),
}

pub fn canonicalize(value: &Value) -> Result<String, CanonicalJsonError> {
    let mut out = String::new();
    encode(value, &mut out)?;
    Ok(out)
}

fn encode(v: &Value, out: &mut String) -> Result<(), CanonicalJsonError> {
    match v {
        Value::Null => out.push_str("null"),
        Value::Bool(b) => out.push_str(if *b { "true" } else { "false" }),
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                write!(out, "{}", i).unwrap();
            } else if let Some(u) = n.as_u64() {
                write!(out, "{}", u).unwrap();
            } else if let Some(f) = n.as_f64() {
                if !f.is_finite() {
                    return Err(CanonicalJsonError::NonFinite(f));
                }
                if f == 0.0 {
                    out.push('0');
                } else if f.fract() == 0.0 && f.abs() < 1e16 {
                    write!(out, "{}", f as i64).unwrap();
                } else {
                    write!(out, "{}", f).unwrap();
                }
            }
        }
        Value::String(s) => write_json_string(&nfc(s), out),
        Value::Array(xs) => {
            out.push('[');
            for (i, x) in xs.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                encode(x, out)?;
            }
            out.push(']');
        }
        Value::Object(map) => {
            out.push('{');
            let mut entries: Vec<(String, &Value)> = map.iter().map(|(k, v)| (nfc(k), v)).collect();
            entries.sort_by(|a, b| a.0.cmp(&b.0));
            for (i, (k, v)) in entries.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                write_json_string(k, out);
                out.push(':');
                encode(v, out)?;
            }
            out.push('}');
        }
    }
    Ok(())
}

fn nfc(s: &str) -> String {
    use unicode_normalization::UnicodeNormalization;
    UnicodeNormalization::nfc(s).collect()
}

fn write_json_string(s: &str, out: &mut String) {
    out.push('"');
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            '\u{08}' => out.push_str("\\b"),
            '\u{0C}' => out.push_str("\\f"),
            c if (c as u32) < 0x20 => {
                write!(out, "\\u{:04x}", c as u32).unwrap();
            }
            c => out.push(c),
        }
    }
    out.push('"');
}

/// Convenience for &serde_json::Map.
pub fn canonicalize_map(map: &Map<String, Value>) -> Result<String, CanonicalJsonError> {
    canonicalize(&Value::Object(map.clone()))
}
